README.md
# Pancake
Lightweight,Fast,And easy-to-use HTML5 game programming library/Game framework for all purposes

Pancake is successor of Cake game engine,Which removed from GitHub

Pancake inspired by LOVE,Pygame,And even AppGameKit and it shares some of their concepts

## Features
- Shares concepts of most lovable frameworks like LOVE,Pygame,And AppGameKit
- Easy and simple to setup
- Free, Moddable, Open-Source,And Cross-Platform
- Fast, Lightweight, And thin in one pack (All in over 40kb, Minimifed version over 30kb)
- Every part needed is written without dependencies,From OS detection to gamepad support
- You can use `CanvasRenderingContext2D` or `WebGLRenderingContext` as graphics core.
- Collision detection physics
- Audio based on WebAudio API
- Video support
- Written in ES5 to support browsers that can't run ES6 (Internet Explorer >= 9 for example)
- Using indexes (Slots), Which can be good for ordering sometimes, And gives easy control of game content
- Optimized for js13kGames game jam

Want to know a lot? See the [documentation](https://github.com/Rabios/Pancake/tree/master/docs/README.md)

## Examples

- Creating a game canvas
```javascript
// Create canvas with both width and height of 600,And set it's index to 0
pancake.canvas.create(600, 600, 0);

// Create context with index 0 and set to use canvas that has index 0
// And set graphics to use context with index 0
pancake.context.create(0, 0);
pancake.graphics.useContext(0);
```

- Drawing text
```javascript
// Create canvas with both width and height of 600,And set it's index to 0
pancake.canvas.create(600, 600, 0);

// Create context with index 0 and set to use canvas that has index 0
// And set graphics to use context with index 0
pancake.context.create(0, 0);
pancake.graphics.useContext(0);

// We can set font and it's size in pixels
pancake.graphics.setFont("Arial", 22);

// Draw text
pancake.graphics.text("Hello Pancake", 10, 10);
```

- Drawing shapes
```javascript
// Create canvas with both width and height of 600,And set it's index to 0
pancake.canvas.create(800, 800, 0);

// Create context with index 0 and set to use canvas that has index 0
// And set graphics to use context with index 0
pancake.context.create(0, 0);
pancake.graphics.useContext(0);

// Draw rectangle with red color
pancake.graphics.color("red");
pancake.graphics.rect(100, 100, 50, 50);

// Draw circle with radius of 50 and orange color
pancake.graphics.color("orange");
pancake.graphics.circle(200, 200, 50);

// Draw line with width of 5 and yellow color
pancake.graphics.color("yellow");
pancake.graphics.line(100, 100, 200, 200, 5);

// Set drawing mode to stroke
pancake.graphics.mode = pancake.graphics.STROKE;

// Draw triangle with green color
pancake.graphics.color("green");
pancake.graphics.triangle(500, 50, 600, 50, 500, 150);

// Draw polygon with points and blue color
pancake.graphics.color("blue");

// From 200, 200 to 200, 300
pancake.graphics.polygon([ [ 200, 200 ],
                           [ 200, 300 ],
                           [ 300, 350 ],
                           [ 200, 200 ] ]);
```

- Playing Audio
```javascript
// When audio loaded, Play it directly
pancake.audio.play("game.mp3");

// Load audio file to index 0
pancake.audio.load("loaded.mp3", 0);

// Plays loaded.mp3
pancake.audio.playFromIndex(0);

// Pauses loaded.mp3
pancake.audio.pauseFromIndex(0);

// Resumes loaded.mp3
pancake.audio.playFromIndex(0);
```

- Saving variables
```javascript
// Save score to localStorage variable "score"
pancake.storage.save("score", Number(0));

// Load it,It gives score value which is 0
var score = Number(pancake.storage.load("score"));
```

- Physics + Game Loop
```javascript
var playercolor = "blue";

// Use a HTML <canvas> tags that has ID "canvas"
// And set index of the context that will use it to 0
pancake.context.use(document.getElementById("canvas"), 0);

// Create 2 rectangles objects
var rect1 = { x: 100, y: 100, w: 50, h: 50 };
var rect2 = { x: 125, y: 125, w: 50, h: 50 };

function game() {
    // Clear canvas
    pancake.graphics.clear();

    // Draw first rectangle
    pancake.graphics.color(playercolor);
    pancake.graphics.rect(rect1.x, rect1.y, rect1.w, rect1.h);
    
    // Draw second rectangle
    pancake.graphics.color("green");
    pancake.graphics.rect(rect2.x, rect2.y, rect2.w, rect2.h);

    // Check collision between both 2 rectangles
    // If true then change color of first rectangle
    // Else then reset it's color to the default color
    if (pancake.physics.checkCollisionRect(rect1.x, rect1.y, rect1.w, rect1.h, rect2.x, rect2.y, rect2.w, rect2.h)) {
        playercolor = "red";
    } else {
        playercolor = "blue";
    }
    
}

var gameloop = pancake.timers.timer(game, 120); // Set frames per second to 120
```

- Keyboard + Game Loop
```javascript
function game() {
    // If key pressed
    if (pancake.input.keypress(pancake.input.key.Z)) alert("Key pressed");

    // If key down
    if (pancake.input.keydown(pancake.input.key.X)) alert("Key down");

    // If key up
    if (pancake.input.keyup(pancake.input.key.C)) alert("Key up");

    // Needed to prevent input loop
    pancake.input.preventLoop();
}

var gameloop = pancake.timers.timer(game, 120); // Set frames per second to 120
```

> NOTES: More examples can be found in the [Examples folder](https://github.com/Rabios/Pancake/tree/master/examples)

## Build

> NOTE 1: Pancake is prebuilt but if you are modify the source then you will need to rebuild by modifying and running build script.

> NOTE 2: As of v0.0.12, I no longer offer build using batch or shell due to being unstable :(

Build using Python:
```batch
mkdir pancake
cd pancake
git clone https://github.com/Rabios/Pancake.git
python build.py canvas
```

> NOTE 3: You can use `webgl` instead of `canvas` when building to generate WebGL version of Pancake!

A folder named `build` will created (if not exist) in the repository folders,Containing `pancake.js`, Which is the full build.

> NOTE 4: Grab the build [here](https://cdn.jsdelivr.net/gh/Rabios/Pancake@master/build/pancake.js),Or [here](https://cdn.jsdelivr.net/gh/Rabios/Pancake@master/build/pancake.min.js) for the minimifed version

## License
```
MIT License

Copyright (c) 2020 - 2021 Rabia Alhaffar

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
```


docs/changelog.md
# Changelog

### v0.0.12
1. Added WebGL renderer that uses `WebGLRenderingContext`!
2. Removed `pancake.timers.second` due to being unstable with intervals and timers.
3. Removed `pancake.graphics.shear` due to lack of usage.
4. Added `pancake.browser.support.WEBGL()` to detect if WebGL supported!
5. Added `pancake.browser.supports(feature)` to make feature detection easier!
6. Due to being batch and shell build unstable, I removed them.

### v0.0.11
1. Added `pancake.game.close()` back so you can close game if opened by script,Or if you use Phonegab or Apache Cordova (In case you linked their script to game HTML file)
2. Optimized the framework for the third time
3. Updated video and audio examples to check if video and audio playing supported
4. Improved browser and OS examples

> CAUTION: If you are developing a game that focuses on frame rates,Do not use `pancake.timers.second` as it's not accurate in counting a second
 
> NOTES: This might be the last update for Pancake,So i maybe add tutorials or more docs if i would,But no updates for the framework source code as it's enough

### v0.0.10
1. Optimized the framework for the second time
2. Fixed example `graphics_04.html` (Changing color mode example),Where color mode doesn't change

> That's all,Thanks for support!

### v0.0.9
1. Optimized the framework for the first time

> TODO: Optimize Pancake for the second time

### v0.0.8
1. Added `pancake.graphics.shadow(color, blur)` to apply shadows in-game when drawing shapes

> TODO: Optimizing

### v0.0.7
1. Removed `pancake.graphics.resize()` but let canvas used by `pancake.canvas.set()` in case you used a canvas using `pancake.context.use()` 
2. Better fullscreen check and better fullscreen functionality
3. Added ability to detect SeaMonkey browser,And Maxthon browser (In case developer didn't changed `navigator.userAgent` from Maxthon browser settings)
4. Improved games,And also improved some examples

> TODO: Optimize Pancake to become smaller,That doesn't means i will remove content

### v0.0.6
1. Added content feature to load JSON content and parse it
2. Added more operating systems and browsers to detect
3. Optimized gamepad detection
4. Added `pancake.graphics.resize()` So you can resize current canvas easily without setting canvas index if it were used from HTML `<canvas>` tags
5. Added `pancake.physics.getDistance()` To get distance from position to another
6. Fixed problem when video and audio not looping if loop set to true
7. Fixed problem mouse pointer not locked in canvas element
8. Fixed lock pointer problems
9. Fixed problem swiping doesn't works

For more info check the [API](https://github.com/Rabios/Pancake/blob/master/docs/api.md)

### v0.0.5
1. Fixed some problems related to touch support
2. Now functions of drawing shapes integrated with `pancake.graphics` instead of adding them to `CanvasRenderingContext2D.prototype`,To let framework take smaller size with same result
3. Removed `pancake.game.close()` (Same as `window.close()` it gives warning)
4. Improved touch
5. Added video features,Check the [API](https://github.com/Rabios/Pancake/blob/master/docs/api.md) again

### v0.0.4
1. Added `pancake.graphics.fullscreen()` to check if game is in fullscreen
2. Set keyboard input to use `KeyboardEvent.keyCode` instead of `KeyboardEvent.key`
3. Fixed problem that some examples can't run offline on internet explorer cause of `localStorage`
   
### v0.0.3
1. Renamed `pancake.timers.animate()` to `window.animate()` cause of TypeError
2. Added new game: Snake 
3. Resize canvas to fit screen when toggling fullscreen mode
4. Added ability use `pancake.graphics.screenshot()` on a specified canvas (you just need to add canvas index as first parameter)
5. Removed `pancake.graphics.setResolution()`

### v0.0.2
1. Removed `pancake.device.online()` cause of incompatibility with Google Chrome and some browsers
2. Improved scrolling
3. Added `pancake.util.randomBetween(a, b)`
4. Added `pancake.timers.second`

### v0.0.1
First release,Looking for bugs to smash it!!!

src/input.js
p.i = {};
p.i.latest_key_down = -1;
p.i.latest_key_up = -1;
p.i.latest_key_pressed = -1;
p.i.latest_mouse_button_down = -1;
p.i.latest_mouse_button_up = -1;
p.i.click = false;
p.i.touchdown = false;
p.i.tap = false;
p.i.mouse_x = 0;
p.i.mouse_y = 0;
p.i.touch_x = 0;
p.i.touch_y = 0;
p.i.touch_start_time = 0;
p.i.swipe_start_x = 0;
p.i.swipe_start_y = 0;
p.i.HORIZONTAL_SWIPE_DIRECTION = "";
p.i.VERTICAL_SWIPE_DIRECTION = "";

w.addEventListener("mousedown", function(e) {
    p.i.swipe_start_x = e.clientX || e.pageX;
    p.i.swipe_start_y = e.clientY || e.pageY;
    p.i.touch_start_time = new Date().getTime();
});

w.addEventListener("mouseup", function(e) {
    p.i.mouse_x = e.clientX || e.pageX;
    p.i.mouse_y = e.clientY || e.pageY;
    var swipe_dist_x = (e.clientX || e.pageX) - p.i.swipe_start_x;
    var swipe_dist_y = (e.clientY || e.pageY) - p.i.swipe_start_y;
    var elapsedTime = new Date().getTime() - p.i.touch_start_time;
    if (elapsedTime <= 1000) {
        if (Math.abs(swipe_dist_x) >= 100 && Math.abs(swipe_dist_y) <= 300) {
            if (swipe_dist_x < 0) p.i.VERTICAL_SWIPE_DIRECTION = "LEFT";
            else p.i.VERTICAL_SWIPE_DIRECTION = "RIGHT";
        }
        else if (Math.abs(swipe_dist_y) >= 100 && Math.abs(swipe_dist_x) <= 300) {
            if (swipe_dist_y < 0) p.i.HORIZONTAL_SWIPE_DIRECTION = "UP";
            else p.i.HORIZONTAL_SWIPE_DIRECTION = "DOWN";
        }
    }
});

w.addEventListener("mousemove",function(e) {
    p.i.mouse_x = e.clientX || e.pageX;
    p.i.mouse_y = e.clientY || e.pageY;
});

w.addEventListener("touchstart", function(e) {
    p.i.touch_x = e.changedTouches[0].clientX || e.changedTouches[0].pageX;
    p.i.touch_y = e.changedTouches[0].clientY || e.changedTouches[0].pageY;
    p.i.swipe_start_x = e.changedTouches[0].pageX;
    p.i.swipe_start_y = e.changedTouches[0].pageY;
    p.i.tap = true;
    p.i.touch_start_time = new Date().getTime();
    e.preventDefault();
}, false);

w.addEventListener("touchend", function(e) {
    p.i.touch_x = e.changedTouches[0].clientX || e.changedTouches[0].pageX;
    p.i.touch_y = e.changedTouches[0].clientY || e.changedTouches[0].pageY;
    var swipe_dist_x = e.changedTouches[0].pageX - p.i.swipe_start_x;
    var swipe_dist_y = e.changedTouches[0].pageY - p.i.swipe_start_y;
    var elapsedTime = new Date().getTime() - p.i.touch_start_time;
    if (elapsedTime <= 1000) {
        if (Math.abs(swipe_dist_x) >= 100 && Math.abs(swipe_dist_y) <= 300) {
            if (swipe_dist_x < 0) p.i.VERTICAL_SWIPE_DIRECTION = "LEFT";
            else p.i.VERTICAL_SWIPE_DIRECTION = "RIGHT";
        }
        else if (Math.abs(swipe_dist_y) >= 100 && Math.abs(swipe_dist_x) <= 300) {
            if (swipe_dist_y < 0) p.i.HORIZONTAL_SWIPE_DIRECTION = "UP";
            else p.i.HORIZONTAL_SWIPE_DIRECTION = "DOWN";
            e.preventDefault();
        }
    }
    p.i.tap = true;
    e.preventDefault();
}, false);

w.addEventListener("touchcancel", function(e) {
    p.i.tap = false;
    e.preventDefault();
}, false);

w.addEventListener("touchmove", function(e) {
    p.i.touch_x = e.changedTouches[0].clientX || e.changedTouches[0].pageX;
    p.i.touch_y = e.changedTouches[0].clientY || e.changedTouches[0].pageY;
    p.i.touchdown = true;
    e.preventDefault();
}, false);

w.addEventListener("click", function() {
    p.i.click = true;
});

w.addEventListener("mousedown", function(e) {
    p.i.latest_mouse_button_down = e.button;
    p.i.click = false;
});

w.addEventListener("mouseup", function(e) {
    p.i.latest_mouse_button_up = e.button;
    p.i.click = false;
});

w.addEventListener("keydown", function(e) {
    p.i.latest_key_down = e.which || e.keyCode;
});

w.addEventListener("keyup", function(e) {
    p.i.latest_key_up = e.which || e.keyCode;
});

w.addEventListener("keypress", function(e) {
    p.i.latest_key_pressed = e.which || e.keyCode;
});

p.i.mousedown = function(b) {
    return p.i.latest_mouse_button_down == b;
};

p.i.mouseup = function(b) {
    return p.i.latest_mouse_button_up == b;
};

p.i.swipe = function(direction) {
    return (p.i.HORIZONTAL_SWIPE_DIRECTION == direction || p.i.VERTICAL_SWIPE_DIRECTION == direction);
};

p.i.keydown = function(k) {
    return p.i.latest_key_down == k;
};

p.i.keyup = function(k) {
    return p.i.latest_key_up == k;
};

p.i.keypress = function(k) {
    return p.i.latest_key_pressed == k;
};

p.i.hideCursor = function(i) {
    pancake.canvases[i].style.cursor = "none";
};

p.i.showCursor = function(i) {
    pancake.canvases[i].style.cursor = "auto";
};

p.i.lockPointer = function() {
    if (p.g.context.canvas.requestPointerLock) p.g.context.canvas.requestPointerLock();
    if (d.pointerLockElement === p.g.context.canvas) p.g.context.canvas.requestPointerLock();
};

p.i.unlockPointer = function() {
    d.exitPointerLock();
};

if (p.b.s.GAMEPAD()) {
    p.i.GAMEPAD_MOVE_HORIZONTAL_DIRECTION = "";
    p.i.GAMEPAD_MOVE_VERTICAL_DIRECTION = "";
    p.i.GAMEPAD_CAMERA_HORIZONTAL_DIRECTION = "";
    p.i.GAMEPAD_CAMERA_VERTICAL_DIRECTION = "";
    p.i.GAMEPAD_ANALOG_UP = -0.3;
    p.i.GAMEPAD_ANALOG_DOWN = 0.3;
    p.i.GAMEPAD_ANALOG_LEFT = 0.3;
    p.i.GAMEPAD_ANALOG_RIGHT = -0.3;
    p.i.GAMEPAD_MOVE_ANALOG = 1;
    p.i.GAMEPAD_CAMERA_ANALOG = 2;

    p.i.gamepadConnected = function(i) {
        return !(n.getGamepads()[i] == undefined);
    };
    
    p.i.gamepadID = function(i) {
        if (!(n.getGamepads()[i] == undefined)) return n.getGamepads()[i].id;
    };
    
    p.i.gamepadButtonPressed = function(b, i) {
        if (!(n.getGamepads()[i] == undefined)) return n.getGamepads()[i].buttons[b].pressed;
    };
    
    p.i.gamepadButtonTouched = function(b, i) {
        if (!(n.getGamepads()[i] == undefined)) return n.getGamepads()[i].buttons[b].touched;
    };

    p.i.checkMovement = function(i, a, d) {
        if (!(n.getGamepads()[i] == undefined)) {
            if (a == p.i.GAMEPAD_MOVE_ANALOG) {
                if (n.getGamepads()[i].axes[1] <= d) p.i.GAMEPAD_MOVE_HORIZONTAL_DIRECTION = "UP";
                if (n.getGamepads()[i].axes[1] >= d) p.i.GAMEPAD_MOVE_HORIZONTAL_DIRECTION = "DOWN";
                if (n.getGamepads()[i].axes[0] <= d) p.i.GAMEPAD_MOVE_VERTICAL_DIRECTION = "LEFT";
                if (n.getGamepads()[i].axes[0] >= d) p.i.GAMEPAD_MOVE_VERTICAL_DIRECTION = "RIGHT";
            }
            
            if (a == p.i.GAMEPAD_CAMERA_ANALOG) {
                if (n.getGamepads()[i].axes[3] <= d) p.i.GAMEPAD_CAMERA_HORIZONTAL_DIRECTION = "UP";
                if (n.getGamepads()[i].axes[3] >= d) p.i.GAMEPAD_CAMERA_HORIZONTAL_DIRECTION = "DOWN";
                if (n.getGamepads()[i].axes[2] <= d) p.i.GAMEPAD_CAMERA_HORIZONTAL_DIRECTION = "LEFT";
                if (n.getGamepads()[i].axes[2] >= d) p.i.GAMEPAD_CAMERA_HORIZONTAL_DIRECTION = "RIGHT";
            }
        }
    };

    p.i.gamepadAnalogMoved = function(i, a, ad, d) {
        p.i.checkMovement(i, a, ad);
        if (a == p.i.GAMEPAD_MOVE_ANALOG) return (p.i.GAMEPAD_MOVE_HORIZONTAL_DIRECTION == d || p.i.GAMEPAD_MOVE_VERTICAL_DIRECTION == d);
        if (a == p.i.GAMEPAD_CAMERA_ANALOG) return (p.i.GAMEPAD_CAMERA_HORIZONTAL_DIRECTION == d || p.i.GAMEPAD_CAMERA_VERTICAL_DIRECTION == d);
    };
}

p.i.preventLoop = function() {
    p.i.latest_key_down = -1;
    p.i.latest_key_up = -1;
    p.i.latest_key_pressed = -1;
    p.i.latest_mouse_button_down = -1;
    p.i.latest_mouse_button_up = -1;
    p.i.click = false;
    p.i.tap = false;
    p.i.touchdown = false;
    p.i.HORIZONTAL_SWIPE_DIRECTION = "";
    p.i.VERTICAL_SWIPE_DIRECTION = "";
    if (p.b.s.GAMEPAD()) {
        p.i.GAMEPAD_MOVE_HORIZONTAL_DIRECTION = "";
        p.i.GAMEPAD_MOVE_VERTICAL_DIRECTION = "";
        p.i.GAMEPAD_CAMERA_HORIZONTAL_DIRECTION = "";
        p.i.GAMEPAD_CAMERA_VERTICAL_DIRECTION = "";
    }
};

// Keyboard keys,Mouse and gamepad buttons
p.i.key = {
    A: 65, B: 66, C: 67, D: 68,E: 69, F: 70, G: 71, H: 72, I: 73, J: 74, K: 75, L: 76, M: 77,
    N: 78, O: 79, P: 80, Q: 81, R: 82, S: 83, T: 84, U: 85, V: 86, W: 87, X: 88, Y: 89, Z: 90,
    ZERO: 48, ONE: 49, TWO: 50, THREE: 51, FOUR: 52, FIVE: 53, SIX: 54, SEVEN: 55, EIGHT: 56,
    NINE: 57, UP: 38, DOWN: 40, LEFT: 37, RIGHT: 39, SPACE: 32, TAB: 9, SHIFT: 16, CONTROL: 17, 
    ALT: 18, BACKSPACE: 8, ENTER: 13, NUMLOCK: 144, OS: 91, UNIDENTIFIED: 0, HOME: 36, PGUP: 33,
    PGDN: 34, CLEAR: 12, DELETE: 46, ESCAPE: 27, INSERT: 45
};

p.i.button = {
    LEFT_MOUSE_BUTTON: 0,RIGHT_MOUSE_BUTTON: 2,MIDDLE_MOUSE_BUTTON: 1,A: 0,B: 1,XBOX_X: 2,
    Y: 3,LB: 4,RB: 5,LT: 6,RT: 7,BACK: 8,START: 9,LEFT_ANALOG_STICK: 10,RIGHT_ANALOG_STICK: 11,
    UP: 12,DOWN: 13,LEFT: 14,RIGHT: 15,PLAYSTATION_X: 0,O: 1,SQUARE: 2,TRIANGLE: 3,L1: 4,R1: 5,L2: 6,R2: 7,
    SELECT: 8
};
